odl/space/fspace.py

```diff
@@ -359,192 +481,416 @@ def element(self, fcall=None, fapply=None):
+        def lincomb_call_in_place(x, out):
+            """Linear combination, in-place version."""
             if not isinstance(out, np.ndarray):
                 raise TypeError('in-place evaluation only possible if output '
                                 'is of type `numpy.ndarray`.')
+            # TODO: this could be optimized for the case when x1 and x2
+            # are identical
             if a == 0 and b == 0:
                 out *= 0
             elif a == 0 and b != 0:
-                y_old_apply(out, *x)
+                x2_call_ip(x, out)
                 if b != 1:
                     out *= b
             elif b == 0 and a != 0:
-                x_old_apply(out, *x)
+                x1_call_ip(x, out)
                 if a != 1:
                     out *= a
             else:
                 tmp = np.empty_like(out)
-                x_old_apply(out, *x)
-                y_old_apply(tmp, *x)
+                x1_call_ip(x, out)
+                x2_call_ip(x, tmp)
                 if a != 1:
                     out *= a
                 if b != 1:
                     tmp *= b
-
                 out += tmp
+            return out
 
-        out._call = lincomb_call
-        out._apply = lincomb_apply
+        out._call_out_of_place = lincomb_call_out_of_place
+        out._call_in_place = lincomb_call_in_place
+        out._call_has_out = out._call_out_optional = True
+        return out
 
-    def zero(self):
-        """The function mapping everything to zero.
+    def _multiply(self, x1, x2, out):
+        """Raw pointwise multiplication of two functions.
 
         Notes
         -----
-        Since `FunctionSpace._lincomb` is slow,
-        we implement this function directly.
-        """
-        def zero_(*_):
-            """The zero function."""
-            return self.field.element(0.0)
-        return self.element(zero_)
-
-    def __eq__(self, other):
-        """Return ``self == other``.
-
-        Returns
-        -------
-        equals : `bool`
-            `True` if ``other`` is a `FunctionSpace` with
-            same `FunctionSet.domain` and `FunctionSet.range`,
-            `False` otherwise.
+        The multiplication is implemented with a simple Python
+        function, so the non-vectorized versions are slow.
         """
-        # TODO: equality also for FunctionSet instances?
-        if other is self:
-            return True
+        # pylint: disable=protected-access
 
-        return (isinstance(other, FunctionSpace) and
-                self.domain == other.domain and
-                self.range == other.range)
+        # Store to allow aliasing
+        x1_call_oop = x1._call_out_of_place
+        x1_call_ip = x1._call_in_place
+        x2_call_oop = x2._call_out_of_place
+        x2_call_ip = x2._call_in_place
+
+        def product_call_out_of_place(x):
+            """The product out-of-place evaluation function."""
+            return x1_call_oop(x) * x2_call_oop(x)
```

Comment:
Skipping one level of wrapping now that I fished out those guys. Could as well be your version.